search/utils.py

```diff
@@ -176,10 +183,21 @@ def search_index(tokens=None, repo_slug=None, sort_by=None, terms=None):
         # Always sort by ID to preserve ordering.
         search = search.sort(sort_by, "id")
 
-    terms = set(get_vocab_names(repo_slug=repo_slug))
-    terms.update(set(('run', 'course', 'resource_type')))
+    terms = set(get_vocab_slugs(repo_slug=repo_slug))
     for term in terms:
-        search.aggs.bucket(term, "terms", field=term)
+        vocab_key = "vocab_{vocab}".format(vocab=term)
+        search.aggs.bucket(
+            "{term}_missing".format(term=vocab_key),
+            "missing", field=vocab_key
+        )
+        search.aggs.bucket(
+            "{term}_buckets".format(term=vocab_key),
+            "terms", field=vocab_key
+        )
+    for term in ('run', 'course', 'resource_type'):
+        search.aggs.bucket(
+            '{term}_builtins'.format(term=term), "terms", field=term
```

Comment:
Appending `_builtins` to prevent conflicts with user-added vocabularies doesn't feel like the right solution. 

However, I don't have a better suggestion right now except for adding another level to the index (turn the dict created by `resource_to_dict` into a nested one with a `vocabularies` key). Then `run`, `course`, and `resource_type` can retain their simple field names in the highest level and the nested dicts will have their own scope.

That's beyond the scope of this PR, but I wanted to make a note. I'll open an issue.